src/modules/peek/Peek.FilePreviewer/Controls/ShellPreviewHandlerControl.xaml.cs

```diff
@@ -0,0 +1,183 @@
+
+        public static readonly DependencyProperty LoadingStateProperty = DependencyProperty.Register(
+            nameof(LoadingState),
+            typeof(PreviewState),
+            typeof(ShellPreviewHandlerControl),
+            new PropertyMetadata(PreviewState.Uninitialized));
+
+        public PreviewState? LoadingState
+        {
+            get { return (PreviewState)GetValue(LoadingStateProperty); }
+            set { SetValue(LoadingStateProperty, value); }
+        }
+
+        public IPreviewHandler? Source
+        {
+            get => _source;
+            set
+            {
+                _source = value;
+                SourcePropertyChanged();
+            }
+        }
+
+        public ShellPreviewHandlerControl()
+        {
+            InitializeComponent();
+
+            containerWndProc = ContainerWndProc;
+        }
+
+        private void SourcePropertyChanged()
+        {
+            if (Source != null)
+            {
+                UpdatePreviewerTheme();
+
+                try
+                {
+                    // Attach the preview handler to the container window
+                    Source.SetWindow(containerHwnd, (RECT*)Unsafe.AsPointer(ref controlRect));
+                    Source.DoPreview();
+
+                    PInvoke.ShowWindow(containerHwnd, SHOW_WINDOW_CMD.SW_SHOW);
+                }
+                catch
+                {
+                }
+            }
+            else
+            {
+                PInvoke.ShowWindow(containerHwnd, SHOW_WINDOW_CMD.SW_HIDE);
+            }
+        }
+
+        private void UpdatePreviewerTheme()
+        {
+            if (Source is IPreviewHandlerVisuals visuals)
+            {
+                try
+                {
+                    switch (ActualTheme)
+                    {
+                        case ElementTheme.Light:
+                            visuals.SetBackgroundColor(new COLORREF(0x00f3f3f3));
+                            visuals.SetTextColor(new COLORREF(0x00000000));
+                            break;
+
+                        case ElementTheme.Dark:
+                            visuals.SetBackgroundColor(new COLORREF(0x00202020));
+                            visuals.SetTextColor(new COLORREF(0x00FFFFFF));
+                            break;
+                    }
+
+                    // Changing the previewer colors might not always redraw itself
+                    PInvoke.InvalidateRect(containerHwnd, (RECT*)null, true);
+                }
+                catch
+                {
+                }
+            }
+        }
+
+        private LRESULT ContainerWndProc(HWND hWnd, uint msg, WPARAM wParam, LPARAM lParam)
+        {
+            // Here for future use :)
+            return PInvoke.DefWindowProc(hWnd, msg, wParam, lParam);
+        }
+
+        private void UserControl_Loaded(object sender, RoutedEventArgs e)
+        {
+            fixed (char* pContainerClassName = "PeekShellPreviewHandlerContainer")
```

Comment:
Does this have to be a fixed statement?

Reply:
It's a weird quirk with CsWin32 that only allows strings as pointers.